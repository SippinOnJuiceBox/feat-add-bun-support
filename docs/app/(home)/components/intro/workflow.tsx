'use client';

import { CheckIcon, Loader2Icon, XIcon } from 'lucide-react';
import { useEffect, useState } from 'react';
import { WorkflowLogs } from './workflow-logs';

type WorkflowLog = {
  duration: number;
  text: string;
};

const Loading = (
  <Loader2Icon
    key="loading"
    className="size-[13px] text-muted-foreground animate-spin"
  />
);
const Success = (
  <div key="success">
    <CheckIcon className="size-[14px] text-emerald-500" />
  </div>
);
const ErrorIndicator = (
  <div key="error">
    <XIcon className="size-[14px] text-rose-500" />
  </div>
);

type LineState = 'idle' | 'loading' | 'success' | 'error';

export const WorkflowExample = ({
  codeBlock,
  logs,
}: {
  codeBlock: React.ReactNode;
  logs: WorkflowLog[];
}) => {
  const [lineStates, setLineStates] = useState<LineState[]>([
    'idle',
    'idle',
    'idle',
  ]);
  const [isRetry, setIsRetry] = useState(false);

  useEffect(() => {
    let cancelled = false;

    const animate = async () => {
      if (isRetry) {
        // On retry with workflow: only the failed step re-runs
        // Steps 1 and 2 are already successful, so skip directly to step 3
        // Wait 2s delay before starting
        await new Promise((resolve) => setTimeout(resolve, 2000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[2] = 'loading';
          return next;
        });

        // Run for 4s (queue + run + success logs)
        await new Promise((resolve) => setTimeout(resolve, 4000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[2] = 'success';
          return next;
        });
      } else {
        // Initial run: reset to idle
        setLineStates(['idle', 'idle', 'idle']);

        // Line 1 (getUser): 3 logs, 3 seconds total
        setLineStates((prev) => {
          const next = [...prev];
          next[0] = 'loading';
          return next;
        });

        await new Promise((resolve) => setTimeout(resolve, 3000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[0] = 'success';
          return next;
        });

        // Line 2 (generateEmail): 2s delay + 3 logs
        await new Promise((resolve) => setTimeout(resolve, 2000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[1] = 'loading';
          return next;
        });

        await new Promise((resolve) => setTimeout(resolve, 3000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[1] = 'success';
          return next;
        });

        // Line 3 (sendEmail): 2s delay + 4 logs, ends in error
        await new Promise((resolve) => setTimeout(resolve, 2000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[2] = 'loading';
          return next;
        });

        await new Promise((resolve) => setTimeout(resolve, 3000));

        if (cancelled) return;

        setLineStates((prev) => {
          const next = [...prev];
          next[2] = 'error';
          return next;
        });

        // Wait 1s after error, then trigger retry
        await new Promise((resolve) => setTimeout(resolve, 1000));

        if (cancelled) return;

        setIsRetry(true);
      }
    };

    animate();

    return () => {
      cancelled = true;
    };
  }, [isRetry]);

  const renderIndicator = (state: LineState) => {
    if (state === 'loading') return Loading;
    if (state === 'success') return Success;
    if (state === 'error') return ErrorIndicator;
    return null;
  };

  return (
    <div className="relative isolate max-w-3xl mx-auto">
      <div className="bg-background border rounded-md overflow-hidden">
        <div className="relative">
          <div className="hidden md:flex absolute z-10 flex-col left-[18px] top-[69px] pointer-events-none select-none">
            {renderIndicator(lineStates[0])}
            <div className="h-[5px]" />
            {renderIndicator(lineStates[1])}
            <div className="h-[46px]" />
            {renderIndicator(lineStates[2])}
          </div>
          {codeBlock}
        </div>
        <WorkflowLogs
          logs={logs}
          className="p-4 font-mono text-sm text-muted-foreground"
        />
      </div>
    </div>
  );
};
