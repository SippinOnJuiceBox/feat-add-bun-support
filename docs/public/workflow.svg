<svg width="1792" height="240" viewBox="0 0 1792 240" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M753.933 55.627C806.249 55.627 840.685 91.718 840.685 147.345C840.685 202.971 806.249 239.062 753.933 239.062C701.286 239.062 666.851 202.971 666.851 147.345C666.851 91.7181 701.286 55.6271 753.933 55.627ZM753.933 90.0625C726.451 90.0626 710.889 111.254 710.889 147.345C710.889 183.436 726.451 204.627 753.933 204.627C781.084 204.627 796.978 183.436 796.978 147.345C796.978 111.254 781.084 90.0625 753.933 90.0625Z" fill="black"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M1445.65 55.627C1497.96 55.627 1532.4 91.718 1532.4 147.345C1532.4 202.971 1497.96 239.062 1445.65 239.062C1393 239.062 1358.56 202.971 1358.56 147.345C1358.56 91.718 1393 55.627 1445.65 55.627ZM1445.65 90.0625C1418.16 90.0625 1402.6 111.254 1402.6 147.345C1402.6 183.436 1418.16 204.627 1445.65 204.627C1472.8 204.627 1488.69 183.436 1488.69 147.345C1488.69 111.254 1472.8 90.0626 1445.65 90.0625Z" fill="black"/>
<path d="M902.473 94.0361C909.095 70.8584 922.34 59.5997 944.855 59.5996H961.079V92.3799H944.524C916.38 92.38 904.128 106.949 904.128 135.094V235.089H861.746V59.5996H901.48L902.473 94.0361Z" fill="black"/>
<path d="M1019.25 138.735L1091.11 59.5996H1143.09L1074.55 132.113L1146.4 235.089H1098.72L1046.74 157.609L1019.25 187.078V235.089H976.872V0H1019.25V138.735Z" fill="black"/>
<path d="M1263.63 32.7803H1235.82C1224.89 32.7803 1217.28 41.3893 1217.28 54.9648V59.5996H1260.98V92.3799H1217.28V235.089H1174.56V92.3799H1149.4V59.5996H1174.56V54.9648C1174.56 19.5359 1191.78 0 1230.85 0H1263.63V32.7803Z" fill="black"/>
<path d="M1322.58 189.396C1322.58 197.673 1327.22 202.309 1335.5 202.31H1348.74V235.089H1324.57C1298.08 235.089 1280.2 218.534 1280.2 191.052V0H1322.58V189.396Z" fill="black"/>
<path d="M1604.59 186.416L1641.02 59.5996H1677.77L1713.86 186.416L1749.95 59.5996H1791.34L1737.7 235.089H1694.32L1659.56 116.883L1624.79 235.089H1581.42L1527.77 59.5996H1568.83L1604.59 186.416Z" fill="black"/>
<path d="M472.173 174.847L514.895 0.671875H554.66L598.038 174.847L640.432 0.671875H685.454L623.015 234H575.692L534.942 77.5713L493.863 234H446.869L384.1 0.671875H429.122L472.173 174.847Z" fill="black"/>
<path d="M125.776 0C120.929 1.64945e-05 117 3.9291 117 8.77539V61.4248C117 66.2711 120.929 70.2002 125.776 70.2002H160.875L161.253 70.2051C169.032 70.4023 175.301 76.6734 175.495 84.4531L175.5 84.8252V143.325C175.5 148.171 179.429 152.101 184.276 152.101H219.375L219.753 152.105C227.53 152.303 233.798 158.571 233.995 166.348L234 166.726V225.225C234 230.071 237.929 234 242.776 234H295.425C300.271 234 304.2 230.071 304.2 225.225V172.575C304.2 167.729 300.271 163.8 295.425 163.8H260.25C252.334 163.76 245.905 157.43 245.705 149.553L245.7 149.176V90.6758C245.7 85.8295 241.771 81.9004 236.925 81.9004H201.825C193.874 81.9004 187.405 75.5551 187.205 67.6523L187.2 67.2754V8.77539C187.2 3.9291 183.271 1.64951e-05 178.425 0H125.776Z" fill="black"/>
<path d="M67.2755 81.9004C62.4293 81.9004 58.5002 85.8295 58.5002 90.6758V143.325C58.5004 148.171 62.4294 152.101 67.2755 152.101H102.375L102.753 152.105C110.53 152.303 116.798 158.571 116.995 166.348L117 166.726V225.225C117 230.071 120.929 234 125.776 234H178.425C183.271 234 187.2 230.071 187.2 225.225V172.575C187.2 167.729 183.271 163.8 178.425 163.8H143.25C135.334 163.76 128.905 157.43 128.705 149.553L128.7 149.176V90.6758C128.7 85.8295 124.771 81.9004 119.925 81.9004H67.2755Z" fill="black"/>
<path d="M8.77454 163.8C3.92844 163.8 0.00012207 167.729 0.00012207 172.575V225.225C0.00012207 230.071 3.92844 234 8.77454 234H61.4249C66.2712 234 70.2003 230.071 70.2003 225.225V172.575C70.2003 167.729 66.2712 163.8 61.4249 163.8H8.77454Z" fill="black"/>
</svg>
