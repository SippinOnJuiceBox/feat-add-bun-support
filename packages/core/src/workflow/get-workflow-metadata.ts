export interface WorkflowMetadata {
  /**
   * Unique identifier for the workflow run.
   */
  workflowRunId: string;

  /**
   * Timestamp when the workflow run started.
   */
  workflowStartedAt: Date;

  /**
   * The URL where the workflow can be triggered.
   */
  url: string;
}

export const WORKFLOW_CONTEXT_SYMBOL =
  /* @__PURE__ */ Symbol.for('WORKFLOW_CONTEXT');

export function getWorkflowMetadata(): WorkflowMetadata {
  // Inside the workflow VM, the context is stored in the globalThis object behind a symbol
  const ctx = (globalThis as any)[WORKFLOW_CONTEXT_SYMBOL] as WorkflowMetadata;
  if (!ctx) {
    throw new Error(
      '`getWorkflowMetadata()` can only be called inside a workflow or step function'
    );
  }
  return ctx;
}
