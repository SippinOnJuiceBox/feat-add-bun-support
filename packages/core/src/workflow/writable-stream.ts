import { STREAM_NAME_SYMBOL, WORKFLOW_GET_STREAM_ID } from '../symbols.js';
import type { WorkflowWritableStreamOptions } from '../writable-stream.js';

export function getWritable<W = any>(
  options: WorkflowWritableStreamOptions = {}
): WritableStream<W> {
  const { namespace } = options;
  const name = (globalThis as any)[WORKFLOW_GET_STREAM_ID](namespace);
  return Object.create(globalThis.WritableStream.prototype, {
    [STREAM_NAME_SYMBOL]: {
      value: name,
      writable: false,
    },
  });
}
