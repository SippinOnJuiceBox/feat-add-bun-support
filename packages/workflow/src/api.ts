export {
  type Event,
  getHookByToken,
  getRun,
  Run,
  resumeHook,
  resumeWebhook,
  runStep,
  type StartOptions,
  start,
  type WorkflowRun,
} from '@workflow/core/runtime';
