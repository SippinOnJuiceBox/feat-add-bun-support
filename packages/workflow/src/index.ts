export * from '@workflow/core';
export * from './stdlib.js';
