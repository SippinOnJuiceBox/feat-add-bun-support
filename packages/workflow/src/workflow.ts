export * from '@workflow/core/_workflow';
export * from './stdlib.js';
