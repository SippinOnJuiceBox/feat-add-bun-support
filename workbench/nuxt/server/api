../../nitro-v2/server/api