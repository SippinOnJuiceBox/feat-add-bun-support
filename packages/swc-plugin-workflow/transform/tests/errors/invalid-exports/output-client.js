import { runStep as __private_run_step } from "workflow/api";
/**__internal_workflows{"steps":{"input.js":{"validStep":{"stepId":"step//input.js//validStep"}}}}*/;
// These should all error - only async functions allowed
export const value = 42;
export function syncFunc() {
    return 'not allowed';
}
export class MyClass {
    method() {}
}
export * from './other';
// This is ok
export async function validStep() {
    return __private_run_step("validStep", {
        arguments: []
    });
}
