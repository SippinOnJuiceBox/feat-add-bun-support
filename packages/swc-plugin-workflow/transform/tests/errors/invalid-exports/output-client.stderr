  x Only async functions can be exported from a "use step" file
   ,-[input.js:4:1]
 3 | // These should all error - only async functions allowed
 4 | export const value = 42;
   : ^^^^^^^^^^^^^^^^^^^^^^^^
 5 | export function syncFunc() {
   `----
  x Only async functions can be exported from a "use step" file
   ,-[input.js:5:1]
 4 |     export const value = 42;
 5 | ,-> export function syncFunc() {
 6 | |     return 'not allowed';
 7 | `-> }
 8 |     export class MyClass {
   `----
  x Functions marked with "use step" must be async functions
   ,-[input.js:5:1]
 4 |     export const value = 42;
 5 | ,-> export function syncFunc() {
 6 | |     return 'not allowed';
 7 | `-> }
 8 |     export class MyClass {
   `----
  x Only async functions can be exported from a "use step" file
    ,-[input.js:8:1]
  7 |     }
  8 | ,-> export class MyClass {
  9 | |     method() {}
 10 | `-> }
 11 |     export * from './other';
    `----
  x Only async functions can be exported from a "use step" file
    ,-[input.js:11:1]
 10 | }
 11 | export * from './other';
    : ^^^^^^^^^^^^^^^^^^^^^^^^
    `----
